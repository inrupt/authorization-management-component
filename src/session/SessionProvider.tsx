// MIT License
//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import type { ReactElement, ReactNode } from "react";
import {
  createContext,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from "react";

import type { Session } from "@inrupt/solid-client-authn-browser";
import {
  fetch,
  getDefaultSession,
  login,
  logout,
} from "@inrupt/solid-client-authn-browser";

import { useRouter } from "next/router";
import { getPodUrlAllFrom, getProfileAll } from "@inrupt/solid-client";
import { getAccessApiEndpoint } from "@inrupt/solid-client-access-grants";
import { getVerifiableCredentialApiConfiguration } from "@inrupt/solid-client-vc";
import type { VerifiableCredentialApiConfiguration } from "@inrupt/solid-client-vc/dist/common/common";
import {
  WebIdNotAvailable,
  DiscoveryNotAvailable,
  AccessGrantsNotSupported,
} from "../../src/errors/customErrors";
import { ProfileModal } from "../components/ProfileModal/ProfileModal";
import { ErrorModal } from "../components/ErrorModal/ErrorModal";

type AuthState =
  | "waiting"
  | "authenticating"
  | "unauthenticated"
  | "authenticated"
  | "pre-authenticating";

export interface ISessionContext {
  login: typeof login;
  logout: typeof logout;
  session: Session;
  fetch: typeof fetch;
  sessionRequestInProgress: boolean;
  state: AuthState;
  accessEndpoint?: string;
  endpointConfiguration?: VerifiableCredentialApiConfiguration;
}

export const createContextBase: () => ISessionContext = () => ({
  login,
  logout,
  fetch,
  session: getDefaultSession(),
  sessionRequestInProgress: true,
  accessEndpoint: undefined,
  state: "waiting",
  endpointConfiguration: undefined,
});

export const SessionContext =
  createContext<ISessionContext>(createContextBase());

/* eslint react/require-default-props: 0 */
export interface ISessionProvider {
  children: ReactNode;
  onError?: (_error: Error) => void;
  restorePreviousSession?: boolean;
}

/**
 * @param session
 * @param callback The callback should mutate react state to trigger re-rendering.
 */
const discoverAccessEndpoint = async (
  session: Session,
  callback: (endpoint: string) => void,
): Promise<void> => {
  if (typeof session.info.webId !== "string") {
    return;
  }
  let profiles;
  try {
    profiles = await getProfileAll(session.info.webId, {
      fetch: session.fetch,
    });
  } catch (e: unknown) {
    throw new WebIdNotAvailable(
      `An error occured when looking up WebID ${session.info.webId}: ${e}`,
    );
  }
  const pods = getPodUrlAllFrom(profiles, session.info.webId);
  if (pods.length < 1) {
    throw new Error(
      `No Pods could be discovered from WebID ${session.info.webId}`,
    );
  }
  // FIXME arbitrarily pick the first Pod.
  const [podRoot] = pods;
  let endpoint;
  try {
    endpoint = await getAccessApiEndpoint(podRoot);
  } catch (e: unknown) {
    throw new DiscoveryNotAvailable(
      `Access endpoint discovery for Pod ${podRoot} failed: ${e}`,
    );
  }
  callback(endpoint);
};

const HintModal = ({
  error,
  modalStatus,
  modalControl,
}: {
  error?: Error;
  modalStatus: boolean;
  modalControl: (status: boolean) => void;
}) => {
  if (error === undefined) {
    return <></>;
  }
  if (error instanceof WebIdNotAvailable) {
    return (
      <ProfileModal
        isOpen={modalStatus}
        onClose={() => {
          modalControl(false);
        }}
      />
    );
  }
  if (error instanceof DiscoveryNotAvailable) {
    return (
      <ErrorModal
        err={
          new AccessGrantsNotSupported(
            "Your Pod Provider doesn't appear to support Access Grants. Please retry using a compliant Pod Provider.",
            {
              cause: error,
            },
          )
        }
        isOpen={modalStatus}
        onClose={() => {
          modalControl(false);
        }}
      />
    );
  }
  return (
    <ErrorModal
      err={error}
      isOpen={modalStatus}
      onClose={() => {
        modalControl(false);
      }}
    />
  );
};

/**
 * Used to provide session data to child components through context.
 */
export default function SessionProvider({
  children,
  onError,
  restorePreviousSession,
}: ISessionProvider): ReactElement {
  const { replace, isReady } = useRouter();
  const session = useMemo(getDefaultSession, []);

  const [state, setState] = useState<AuthState>(
    session.info.isLoggedIn ? "authenticated" : "waiting",
  );

  const [accessEndpoint, setAccessEndpoint] = useState<string>();
  const [accessEndpointConfig, setAccessEndpointConfig] =
    useState<VerifiableCredentialApiConfiguration>();
  const [modalOpen, setModalOpen] = useState<boolean>(true);
  const [errorState, setErrorState] = useState<Error>();

  const isAuthenticated = state === "authenticated";
  const isUnAuthenticated = state === "unauthenticated";

  useEffect(() => {
    if (isReady) {
      if (state === "waiting") {
        setState("authenticating");
      } else if (state === "authenticating") {
        session
          .handleIncomingRedirect({
            url: window.location.href,
            restorePreviousSession,
          })
          .then((res) => {
            setState(res?.isLoggedIn ? "authenticated" : "unauthenticated");
          })
          .catch((error) => {
            if (onError) {
              onError(error as Error);
            } else {
              throw error;
            }
            setState(
              session.info?.isLoggedIn ? "authenticated" : "unauthenticated",
            );
          });
      }
    }
  }, [state, isReady, onError, replace, session, restorePreviousSession]);

  useEffect(() => {
    session.events.on("logout", () => {
      if (isAuthenticated || isUnAuthenticated) {
        setState("waiting");
      }
    });
  }, []);

  useEffect(() => {
    if (state === "authenticated") {
      discoverAccessEndpoint(session, (endpoint: string) => {
        setAccessEndpoint(endpoint);
      }).catch((e) => {
        if (e instanceof Error) {
          setErrorState(e);
        } else {
          console.error(`An error occured: ${e}`);
        }
      });
    }
  }, [state]);

  useEffect(() => {
    if (typeof accessEndpoint === "string") {
      getVerifiableCredentialApiConfiguration(accessEndpoint).then(
        setAccessEndpointConfig,
        (error: unknown) => {
          if (error instanceof Error) {
            setErrorState(error);
          } else {
            console.error(`An error occured: ${error}`);
          }
        },
      );
    }
  }, [accessEndpoint]);

  const loginCallback = useCallback(
    async (options: Parameters<Session["login"]>[0]) => {
      if (isUnAuthenticated) {
        setState("pre-authenticating");
        try {
          await session.login(options);
        } catch (error) {
          if (onError) {
            onError(error as Error);
          } else {
            throw error;
          }
        } finally {
          setState("unauthenticated");
        }
      }
    },
    [isUnAuthenticated, onError, session],
  );

  const logoutCallback = useCallback(
    async (options: Parameters<Session["logout"]>[0]) => {
      if (isAuthenticated) {
        try {
          await session.logout(options);
          setState("unauthenticated");
        } catch (error) {
          if (onError) {
            onError(error as Error);
          } else {
            throw error;
          }
        }
      }
    },
    [isAuthenticated, onError, session],
  );

  const value = useMemo<ISessionContext>(
    () => ({
      login: loginCallback,
      logout: logoutCallback,
      fetch: session.fetch,
      session,
      sessionRequestInProgress:
        state === "waiting" || state === "authenticating",
      state,
      accessEndpoint,
      endpointConfiguration: accessEndpointConfig,
    }),
    [
      loginCallback,
      logoutCallback,
      state,
      session,
      accessEndpoint,
      accessEndpointConfig,
    ],
  );

  return (
    <SessionContext.Provider value={value}>
      <>
        <HintModal
          modalControl={setModalOpen}
          modalStatus={modalOpen}
          error={errorState}
        ></HintModal>
        {children}
      </>
    </SessionContext.Provider>
  );
}
