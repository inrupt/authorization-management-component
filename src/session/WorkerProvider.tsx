//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
import {
  ReactElement,
  ReactNode,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";

import { v4 } from "uuid";
import { isValidAccessGrant as _isValidAccessGrant } from "@inrupt/solid-client-access-grants";
import {
  getResourceInfo,
  getContentType as solidClientGetContentType,
} from "@inrupt/solid-client";
import { PurposeInfo, PurposeRequest, PurposeResponse } from "./workertypes";
import { SessionContext } from "./SessionProvider";
import {
  getFromWebIdHelper,
  namePredicates,
  imagePredicates,
} from "../helpers/profile/profile";

export interface IWorkerContext {
  getPurposeInfo: (_url: string) => Promise<PurposeInfo>;
  getNameFromWebId: (_url: string) => Promise<string | null>;
  getImageFromWebId: (_url: string) => Promise<string | null>;
  getContentType: (_url: string) => Promise<string | null>;
  isValidAccessGrant: (
    ..._args: Parameters<typeof _isValidAccessGrant>
  ) => Promise<boolean>;
  nameRecord: Record<string, string>;
}

export const WorkerContext = createContext<IWorkerContext>({
  getPurposeInfo: () =>
    new Promise<PurposeInfo>(() => {
      /* promise never resolves */
    }),
  getNameFromWebId: () =>
    new Promise<string | null>(() => {
      /* promise never resolves */
    }),
  getImageFromWebId: () =>
    new Promise<string | null>(() => {
      /* promise never resolves */
    }),
  getContentType: () =>
    new Promise<string | null>(() => {
      /* promise never resolves */
    }),
  isValidAccessGrant: async (vc, options) => {
    const valid = await _isValidAccessGrant(vc, options);
    return valid.errors.length === 0;
  },
  nameRecord: {},
});

/* eslint react/require-default-props: 0 */
export interface IWorkerProvider {
  children: ReactNode;
}

/**
 * Used to provide session data to child components through context.
 */
export default function WorkerProvider({
  children,
}: IWorkerProvider): ReactElement {
  const { fetch } = useContext(SessionContext);
  const [worker, setWorker] = useState<Worker>();
  const nameRecord = useMemo<Record<string, Promise<string | null>>>(
    () => ({}),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [fetch]
  );
  const imageRecord = useMemo<Record<string, Promise<string | null>>>(
    () => ({}),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [fetch]
  );
  const purposeRecord = useMemo<Record<string, PurposeInfo>>(() => ({}), []);
  const invalidatedUrl = useMemo<Set<string>>(() => new Set(), []);
  const [resolvedNameRecord, setResolvedNameRecord] = useState<
    Record<string, string>
  >({});

  useEffect(() => {
    setWorker(new Worker(new URL("./worker.ts", import.meta.url)));
  }, []);

  const getImageFromWebId = useCallback(
    // eslint-disable-next-line no-return-assign
    async (url: string) =>
      (imageRecord[url] ??= getFromWebIdHelper(url, imagePredicates, {
        fetch,
        allowedNamed: true,
      })),
    [fetch, imageRecord]
  );

  const getNameFromWebId = useCallback(
    // eslint-disable-next-line no-return-assign
    async (url: string) =>
      (nameRecord[url] ??= getFromWebIdHelper(url, namePredicates, {
        fetch,
      })).then((name) => {
        if (name && !(name in resolvedNameRecord)) {
          setResolvedNameRecord({ ...resolvedNameRecord, [url]: name });
        }
        return name;
      }),
    // We don't want to make resolvedNameRecord a dependency of the callback
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [fetch, nameRecord]
  );

  const getContentType = useMemo(() => {
    const cache: Record<string, Promise<string | null>> = {};
    return async (resourceIri: string) => {
      if (!(resourceIri in cache)) {
        cache[resourceIri] = getResourceInfo(resourceIri, {
          fetch,
        })
          .then((c) => solidClientGetContentType(c))
          .catch(() => {
            delete cache[resourceIri];
            return null;
          });
      }
      return cache[resourceIri];
    };
  }, [fetch]);

  const isValidAccessGrant = useCallback(
    async (...args: Parameters<typeof _isValidAccessGrant>) => {
      let [vcUrl] = args;

      if (
        typeof vcUrl === "object" &&
        !(vcUrl instanceof URL) &&
        "id" in vcUrl
      ) {
        vcUrl = vcUrl.id;
      }

      if (invalidatedUrl.has(vcUrl.toString())) {
        return false;
      }

      const result = await _isValidAccessGrant(...args);
      if (result.errors.length !== 0) {
        invalidatedUrl.add(vcUrl.toString());
        return false;
      }
      return true;
    },
    // We don't want to make resolvedNameRecord a dependency of the callback
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  const getPurposeInfo = useCallback(
    async (url: string) => {
      if (url in purposeRecord) {
        return purposeRecord[url];
      }

      return new Promise<PurposeInfo>((res) => {
        const fixedWorker = worker;
        const id = v4();

        const listener = ({ data }: MessageEvent) => {
          const result: PurposeResponse = JSON.parse(data);
          if (result.id === id) {
            fixedWorker?.removeEventListener("message", listener);
            purposeRecord[url] = result.value;
            res(result.value);
          }
        };
        fixedWorker?.addEventListener("message", listener);
        fixedWorker?.postMessage(
          JSON.stringify({
            type: "purpose",
            value: url,
            id,
          } as PurposeRequest)
        );
      });
    },
    [worker, purposeRecord]
  );

  const value = useMemo(
    () => ({
      getPurposeInfo,
      getNameFromWebId,
      getImageFromWebId,
      isValidAccessGrant,
      nameRecord: resolvedNameRecord,
      getContentType,
    }),
    [
      getPurposeInfo,
      getNameFromWebId,
      getImageFromWebId,
      resolvedNameRecord,
      isValidAccessGrant,
      getContentType,
    ]
  );

  return (
    <WorkerContext.Provider value={value}>{children}</WorkerContext.Provider>
  );
}
