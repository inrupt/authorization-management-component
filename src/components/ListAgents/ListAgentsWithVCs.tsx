//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import { Session } from "@inrupt/solid-client-authn-browser";
import { useContext, useEffect, useState } from "react";
import { isValidAccessGrant } from "@inrupt/solid-client-access-grants";
import { VerifiableCredentialApiConfiguration } from "@inrupt/solid-client-vc/dist/common/common";
import getRequestors from "../../helpers/requestors/requestors";
import { SessionContext } from "../../session/SessionProvider";
import Loader from "../Loader/Loader";
import ListAgents from "./ListAgents";
import { WorkerContext } from "../../session/WorkerProvider";

async function getAgents(
  accessEndpoint: string | undefined,
  endpointConfig: VerifiableCredentialApiConfiguration | undefined,
  session: Session,
  _isValidAccessGrant: (
    ..._args: Parameters<typeof isValidAccessGrant>
  ) => Promise<boolean>
) {
  const { webId } = session.info;
  const { fetch } = session;

  if (
    accessEndpoint === undefined ||
    endpointConfig === undefined ||
    typeof webId !== "string"
  ) {
    return [];
  }

  return [
    ...(await getRequestors({
      fetch,
      webId,
      accessEndpoint,
      isValidAccessGrant: _isValidAccessGrant,
      endpointConfiguration: endpointConfig,
    })),
  ];
}

function SuspendedList({
  onDetails,
  searchString,
  agents,
}: {
  onDetails: (_agent: string) => void;
  searchString: string;
  agents?: string[];
}) {
  if (agents === undefined) {
    return <Loader />;
  }
  return (
    <ListAgents
      agents={agents}
      onDetails={onDetails}
      searchString={searchString}
    />
  );
}

export default function ListAgentsWithVCs({
  onDetails,
  searchString,
  onNotEmptyAgents,
}: {
  onDetails: (_agent: string) => void;
  searchString: string;
  onNotEmptyAgents?: () => void;
}) {
  const { session, accessEndpoint, endpointConfiguration } =
    useContext(SessionContext);
  const { isValidAccessGrant } = useContext(WorkerContext);
  const [agents, setAgents] = useState<string[] | undefined>([]);

  useEffect(() => {
    if (accessEndpoint !== undefined && endpointConfiguration !== undefined) {
      // Use undefined value for agent list as a proxy for loading state.
      setAgents(undefined);
      getAgents(
        accessEndpoint,
        endpointConfiguration,
        session,
        isValidAccessGrant
      )
        .then(setAgents)
        .catch((error) => {
          console.error(error);
          setAgents([]);
        });
    }
  }, [accessEndpoint, endpointConfiguration, session, isValidAccessGrant]);

  useEffect(() => {
    if (agents !== undefined && agents.length > 0) {
      if (onNotEmptyAgents) {
        onNotEmptyAgents();
      }
    }
  }, [agents]);

  return (
    <SuspendedList
      agents={agents}
      onDetails={onDetails}
      searchString={searchString}
    />
  );
}
