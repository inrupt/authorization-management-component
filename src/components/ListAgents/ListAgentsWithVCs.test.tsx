// MIT License
//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//

import { render } from "@testing-library/react";
import ListAgentsWithVCs from "./ListAgentsWithVCs";
import WorkerProvider from "../../session/WorkerProvider.mock";
import SessionProvider from "../../session/SessionProvider.mock";

describe("ListAgents", () => {
  it("defaults to an empty list", async () => {
    const result = render(
      <SessionProvider mockAccessEndpoint={undefined}>
        <WorkerProvider>
          <ListAgentsWithVCs
            onDetails={jest.fn()}
            searchString="mock search string"
          />
        </WorkerProvider>
      </SessionProvider>,
    );
    expect(result.baseElement).toMatchSnapshot();
  });
});
