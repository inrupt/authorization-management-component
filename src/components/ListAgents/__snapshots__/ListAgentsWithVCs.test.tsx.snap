// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListAgents defaults to an empty list 1`] = `
<body>
  <div>
    <div
      class="empty-grant-container"
    >
      <i
        class="bi bi-inboxes-fill empty-inbox"
      />
      <p>
        <strong>
          You have not yet granted access to anyone.
        </strong>
      </p>
      <p>
        Once you have granted access, you will be able to see it on this page.
      </p>
    </div>
  </div>
</body>
`;
